classes/robot/crawler.php

```diff
@@ -567,6 +568,7 @@ public function process_queue($verbose = false) {
                 $persistent = new url(0, $node);
                 $persistent->update();
                 $lock->release();
+                $hastime = time() < $cronstop; // Check if we need to break from processing
```

Comment:
I suspect the flawed logic might be in this loop specifically.

since when stracing they seemed to be spinning calling update and then checking the lock advisory table.

Reply:
Perhaps it could be specifically courses that have a lastcrawled of null, they are always eligible for queue selection.

And setting the needscrawl to lastcrawled will simply set it to null, which means they will always be eligible for queue selection even if the course is never in recent courses, and this would be an infinite loop

Reply:
and if all the available cron processes are tied up running this loop, there will never be time/space to update the needscrawl field.

Reply:
perhaps we should skip the recentcourses check if lastcrawled is null?